src/TorchSharp/JIT/ScriptModule.cs

```diff
@@ -268,91 +269,102 @@ private Type GetType(Type type)
 #endif
 
                 [DllImport("LibTorchSharp")]
-                private static extern IntPtr THSJIT_Module_forward(HType module, IntPtr tensors, int length);
+                private static extern void THSJIT_Module_forward(HType module, IntPtr tensors, int length, AllocatePinnedArray allocator, out sbyte typeCode);
 
-                /// <summary>
-                /// Invoke the 'forward' function of the script with one tensor as its argument
-                /// </summary>
-                /// <param name="tensor">The input tensor</param>
-                /// <returns></returns>
-                public unsafe override Tensor forward(Tensor tensor)
+                public object forward(params object[] objs)
                 {
-                    var tensorRefs = stackalloc[] { tensor.Handle };
-                    var res = THSJIT_Module_forward(handle, (IntPtr)tensorRefs, 1);
-                    if (res == IntPtr.Zero)
-                        CheckForErrors();
-                    return new Tensor(res);
+                    if (!objs.All(o => typeof(Tensor).IsAssignableFrom(o.GetType()))) {
+                        throw new NotImplementedException("ScriptModule.forward() taking non-tensors as input arguments");
+                    }
+
+                    IntPtr[] ptrArray = null;
+                    sbyte typeCode = 0;
+
+                    using (var parray = new PinnedArray<IntPtr>()) {
+
+                        var tensors = objs.Select(o => (Tensor)o).ToArray();
+                        var count = tensors.Length;
+                        var tensorRefs = new IntPtr[count];
+                        for (var i = 0; i < tensors.Length; i++) tensorRefs[i] = tensors[i].Handle;
+
+                        THSJIT_Module_forward(handle, parray.CreateArray(tensorRefs), count, parray.CreateArray, out typeCode);
+                        torch.CheckForErrors();
+                        ptrArray = parray.Array;
+                    }
+
+
+                    switch (typeCode) {
+                    case 0:
+                        // Nothing.
+                        throw new NotImplementedException("ScriptModule.forward() returning something else than a tensor, a tuple of tensors, or list of tensors.");
+                    case 1:
+                        // Tensor
+                        return new Tensor(ptrArray[0]);
+                    case 2:
```

Comment:
Can C++ modules return a tuple of 0 elements?

Reply:
I don't think so. If we encounter that, we can revisit.